executor/studentapi.pyx

```diff
@@ -0,0 +1,238 @@
+
+    `CodeExecutionError` accepts arbitrary data that can be examined or written into logs.
+
+    Example:
+        >>> err = CodeExecutionError('Error', dev_ids=[2, 3])
+        >>> err
+        CodeExecutionError('Error', dev_ids=[2, 3])
+
+    """
+
+    def __init__(self, message='', **context):
+        super().__init__(message)
+        self.context = context
+        # log_runtime(ERROR, str(self).encode('utf-8'))
+
+    def __repr__(self):
+        cls_name, (msg, *_) = self.__class__.__name__, self.args
+        kwargs = ', '.join(f'{name}={repr(value)}' for name, value in self.context.items())
+        return f'{cls_name}({repr(msg)} {kwargs})'
+
+
+cdef class Gamepad:
+    """
+    The API for accessing gamepads.
+
+    Attributes:
+        mode: The execution state of the robot.
+    """
+    cdef public str mode
+
+    def __cinit__(self, mode):
+        """Initializes the mode of the robot. Also initializes the auxiliary SHM. """
+        self.mode = mode
+        # shm_aux_init(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM initialized")
+
+    def __dealloc__(self):
+        """Once process is finished and object is deallocated, close the mapping to the auxiliary SHM."""
+        # shm_aux_stop(STUDENTAPI)
+        # log_runtime(DEBUG, "Aux SHM stopped")
+
+    cpdef get_value(self, str param_name):
+        """
+        Get a gamepad parameter if the robot is in teleop.
+
+        Args:
+            param: The name of the parameter to read. Possible values are at https://pioneers.berkeley.edu/software/robot_api.html 
+        """
+        if self.mode != 'teleop':
+            raise CodeExecutionError(f'Cannot use Gamepad during {self.mode}')
+        cdef bytes param = param_name.encode('utf-8')
+        cdef uint32_t buttons
+        cdef float joysticks[4]
+        cdef char** button_names 
+        cdef char** joystick_names 
+        with nogil:
+            gamepad_read(&buttons, joysticks)
+            button_names = get_button_names()
+            joystick_names = get_joystick_names()
+        for i in range(NUM_GAMEPAD_BUTTONS):
+            if param == button_names[i]:
+                return bool(buttons & (1 << i))
+        for i in range(4):
+            if param == joystick_names[i]:
+                return joysticks[i]
+
+
+cdef class Robot:
+    """
+    The API for accessing the robot and its devices.
+    """
+    # Include device handler and executor API
+    cdef dict running_actions
+
+    def __cinit__(self):
+        """Initializes the shared memory (SHM) wrapper. """
+        shm_init(STUDENTAPI) # Remove once integrated into executor?
+        log_runtime(DEBUG, "SHM intialized")
+        self.running_actions = {}
+
+    def __dealloc__(self):
+        """Once process is done and object is deallocated, close the mapping to SHM."""
+        shm_stop(STUDENTAPI)
+        log_runtime(DEBUG, "SHM stopped")
+
+
+    def run(self, action, *args, **kwargs) -> None:
```

Comment:
FYI, some subtleties with actions:
1. [`Robot.run`](https://pioneers.berkeley.edu/software/robot_api.html#robot-run-async-function-args-) is supposed to accept a coroutine function (defined with `async def`), so accepting regular functions for `action` would break existing student code. Threads are not a drop-in replacement for [awaitables](https://docs.python.org/3/library/asyncio-task.html#awaitables).
1. There's no guarantee a student-supplied function will terminate quickly, if at all. In fact, the rationale for actions is to support [long-running time-sensitive tasks](https://pioneers.berkeley.edu/software/auto.html) outside of the `setup/main` loop. Counting on actions to terminate eventually is not a good solution, IMO.
1. For device contention reasons, an action should only have one copy running at a [given time](https://github.com/pioneers/PieCentral/blob/master/runtime/runtime/studentapi.py#L206).

The original action executor implementation was carefully designed to avoid these problems. A dedicated thread runs student actions using `asyncio`, and the actions are scheduled from the main thread executing `setup/main`. Unlike threads, awaitables can be safely cancelled.

Reply:
1. Yea I know this. I thought it would simplify the student code in the future though it is backward incompatible. I also think its better if the students don't have to worry about using `async` at all, if its possible.
2. Yea I agree. I'm thinking of now creating a flag/exception wrapper around the student's function so that I can cancel the actions safely without doing `pthread_cancel`. I have a similar issue with the main thread to safely cancel between auton and teleop that I still need to figure out.
3. I didn't think about this, but I can change it so that if the action is currently running it will throw an error/warning. What behavior do you think should happen when the student schedules multiple actions of the same name?

Reply:
1. I agree that `async`/`await` syntax has historically caused a lot of confusion for students and within PiE, but this can be overcome with better documentation. `asyncio` offers several [advantages](https://discuss.python.org/t/what-are-the-advantages-of-asyncio-over-threads/2112/6) over manual thread management for our use case (especially cancellation, see below).
1. Student code is completely opaque, so using `threading.Event` or `queue.Queue` to pass halt messages in the standard way is not possible without a massively complex API change. If we insist on synchronous actions, [a thread pool](https://docs.python.org/3/library/concurrent.futures.html) caps the number of workers, but even a pool cannot terminate tasks when going back to idle (see the note on `shutdown`). The executor process would need to exit and be respawned.
1. If the action is already running, simply log the attempt at `DEBUG`. It's not anomalous to schedule an action multiple times, since we allow `Robot.run` to be called from `(autonomous|teleop)_main`, which is called very frequently. `Robot.run`'s semantics follow "run, if not already running". [Staff Curry](https://github.com/pioneers/PieCentral/blob/master/staff-bots/staff_curry_code.py#L67-L84) illustrates this well: the button for waving might be pressed for many consecutive cycles, but obviously it's all part of one wave. The unfortunate reality is that very few people remember to use `Robot.is_running(action)`.

The original [API design notes](https://drive.google.com/file/d/0B0UMIFiWC3aJZnF1WVdfZTd3b1U/view?usp=sharing), in case you find them helpful.

Reply:
2. I think I figured a way to cancel the Python threads safely. It uses `PyThreadState_SetAsyncExc`, which is a Python C API function, to send an exception to a thread. You can see the code for this in the `ThreadWrapper` in `studentapi.pyx` and in `actions_handler` in `executor.c`. I do agree though that `asyncio` is probably safer and perhaps faster due to OS overhead. I might implement that later instead of `threading` if `threading` has issues, but at least in my testing, it seems to work.
3. Changed the function to these semantics.